### PR TITLE
DS-3759 Implement the patch method for the workspaceitems endpoint

### PR DESCRIPTION
**Pull request by abollini**
This PR addresses the following issues
- list existent submissions (DS-3743)
- create new empty submission (DS-3740)
- update (patch) existent submissions (DS-3759)

Please check the REST contract proposed here: https://github.com/DSpace/Rest7Contract/pull/16
further documentation about the additional configuration files introduced is available here:
https://wiki.duraspace.org/display/DSPACE/Configuration+changes+in+the+submission+process

### CODE REVIEW

**Comment by abollini** [comment]
@tdonohue  @tomdesair and anyone else interested

The PR is flagged as workinprogress for two reasons.
1. I'm open to proposal about the best way to review it as it is quite large. So I have created intermediate PRs (see the description) if this is preferred
2. we are still improving the code to exactly meet the REST contract that is presented here: https://github.com/DSpace/Rest7Contract/pull/16

Other than the documentation in the above rest contract PR I have updated the wiki page about the configuration changes describing the new introduced file, see https://wiki.duraspace.org/display/DSPACE/Configuration+changes+in+the+submission+process

Nevertheless, I will appreciate an ongoing review

File: dspace-api/src/main/java/org/dspace/app/itemupdate/AddBitstreamsAction.java [hunk 1]
```diff
@@ -210,7 +210,7 @@ protected String addBitstream(Context context, ItemArchive itarch, Item item, Fi
 				if (group != null)
 				{
                     authorizeService.removeAllPolicies(context, bs);  // remove the default policy
-                    authorizeService.createResourcePolicy(context, bs, group, null, ce.permissionsActionId, null);
+                    authorizeService.createResourcePolicy(context, bs, group, null, ce.permissionsActionId, null, null, null, null, null);
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/itemupdate/AddBitstreamsAction.java]: This call looks very odd (with 5 `null` values passed in as the last five attributes).  I wonder if we should simply have multiple `createResourcePolicy()` methods with different numbers of inputs, so that we don't have to pass in such large numbers of `null` inputs.

File: dspace-api/src/main/java/org/dspace/submit/step/AccessStep.java [hunk 2]
```diff
@@ -7,318 +7,28 @@
...
-    @Override
-    public int getNumberOfPages(HttpServletRequest request,
-            SubmissionInfo subInfo) throws ServletException
-    {
-        return 1;
-
-    }
+	@Override
+	public void doProcessing(Context context, Request req, InProgressSubmission wsi) {
+		// TODO Auto-generated method stub
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/submit/step/AccessStep.java]: I'd like to see a more informative comment here. There's several areas of this PR which have this exact same `// TODO Auto-generated method stub` comment.  If these methods should be empty, let's add in a comment that says why they are empty, instead of having a `TODO` statement.

Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/submit/step/AccessStep.java]: Also, it is unclear to me why several of these classes are being kept with entirely empty methods.  Are these classes (such as this `AccessStep` one) now deprecated?  (If so, they should be flagged as such)  Is there some other reason they are not just deleted?

File: dspace-api/src/main/java/org/dspace/workflowbasic/BasicWorkflowItemServiceImpl.java [hunk 3]
```diff
@@ -161,4 +162,11 @@ public void update(Context context, BasicWorkflowItem workflowItem) throws SQLEx
     public int countTotal(Context context) throws SQLException {
         return workflowItemDAO.countRows(context);
     }
+    
+	@Override
+	public void move(Context context, BasicWorkflowItem inProgressSubmission, Collection fromCollection,
+			Collection toCollection) {
+		// TODO not implemented yet
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/workflowbasic/BasicWorkflowItemServiceImpl.java]: Is this purposefully not implemented?  Or will this be coming in the PR?  It'd be good to clarify this `TODO` comment.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java [hunk 4]
```diff
@@ -26,46 +26,48 @@
...
+ * This is a specialized controller to provide access to the bitstream binary
+ * content
  * 
  * @author Andrea Bollini (andrea.bollini at 4science.it)
  *
  */
 @RestController
-@RequestMapping("/api/"+BitstreamRest.CATEGORY +"/"+ BitstreamRest.PLURAL_NAME + "/{uuid:[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}}/content")
+@RequestMapping("/api/" + BitstreamRest.CATEGORY + "/" + BitstreamRest.PLURAL_NAME
+		+ "/{uuid:[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}}/content")
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/BitstreamContentRestController.java]: Please add a comment explaining this very large regex. It is not self-explanatory.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java [hunk 5]
```diff
@@ -92,30 +117,30 @@ public void afterPropertiesSet() {
...
 	@RequestMapping(method = RequestMethod.GET, value = "/{id:\\d+}")
 	@SuppressWarnings("unchecked")
-	public DSpaceResource<RestModel> findOne(@PathVariable String apiCategory, @PathVariable String model,
+	public DSpaceResource<DirectlyAddressableRestModel> findOne(@PathVariable String apiCategory, @PathVariable String model,
 			@PathVariable Integer id, @RequestParam(required = false) String projection) {
 		return findOneInternal(apiCategory, model, id, projection);
 	}
 
-	@RequestMapping(method = RequestMethod.GET, value = "/{id:[A-z0-9]+}")
+	@RequestMapping(method = RequestMethod.GET, value = "/{id:^(?!^\\d+$)(?!^[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}$)[\\w+\\-]+$+}")
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java]: Please add a comment explaining this very large regex. It is not self-explanatory.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java [hunk 6]
```diff
@@ -134,7 +159,7 @@ public ResourceSupport findRel(HttpServletRequest request, @PathVariable String
 		return findRelInternal(request, apiCategory, model, id, rel, page, assembler, projection);
 	}
 
-	@RequestMapping(method = RequestMethod.GET, value = "/{id:[A-z0-9]+}/{rel}")
+	@RequestMapping(method = RequestMethod.GET, value = "/{id:^(?!^\\d+$)(?!^[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}$)[\\w+\\-]+$}/{rel}")
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java]: Please add a comment explaining this very large regex. It is not self-explanatory.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java [hunk 7]
```diff
@@ -148,18 +173,115 @@ public ResourceSupport findRel(HttpServletRequest request, @PathVariable String
 		return findRelInternal(request, apiCategory, model, uuid, rel, page, assembler, projection);
 	}
 
-	@RequestMapping(method = RequestMethod.GET, value = "/{id:[A-z0-9]+}/{rel}/{relid:[A-z0-9]+}")
+	@RequestMapping(method = RequestMethod.GET, value = "/{id:^(?!^\\d+$)(?!^[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}$)[\\w+\\-]+$}/{rel}/{relid:[\\w+\\-]+}")
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java]: Please add a comment explaining this very large regex. It is not self-explanatory.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java [hunk 8]
```diff
@@ -148,18 +173,115 @@ public ResourceSupport findRel(HttpServletRequest request, @PathVariable String
...
+
+	@RequestMapping(method = RequestMethod.POST, value = "/{id:\\d+}" , headers = "content-type=multipart/form-data")
+	public <ID extends Serializable> ResponseEntity<ResourceSupport> upload(HttpServletRequest request,
+			@PathVariable String apiCategory, @PathVariable String model, @PathVariable Integer id,
+			@RequestParam(required=false, value="extraField") String extraField,
+		    @RequestParam("file") MultipartFile uploadfile) throws HttpRequestMethodNotSupportedException {
+		return uploadInternal(request, apiCategory, model, id, extraField, uploadfile);
+	}
+	
+	@RequestMapping(method = RequestMethod.POST, value = "/{uuid:[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}}", headers = "content-type=multipart/form-data")
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java]: Please add a comment explaining this very large regex. It is not self-explanatory.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java [hunk 9]
```diff
@@ -148,18 +173,115 @@ public ResourceSupport findRel(HttpServletRequest request, @PathVariable String
...
+		return ControllerUtils.toEmptyResponse(HttpStatus.INTERNAL_SERVER_ERROR);
+	}	
+	
+	@RequestMapping(method = RequestMethod.PATCH, value = "/{id:\\d+}")
+	public ResponseEntity<ResourceSupport> patch(HttpServletRequest request, @PathVariable String apiCategory,
+			@PathVariable String model, @PathVariable Integer id, @RequestBody(required = true) JsonNode jsonNode) throws HttpRequestMethodNotSupportedException {
+		return patchInternal(request, apiCategory, model, id, jsonNode);
+	}
+
+	@RequestMapping(method = RequestMethod.PATCH, value = "/{uuid:[0-9a-fxA-FX]{8}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{4}-[0-9a-fxA-FX]{12}}")
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/RestResourceController.java]: Please add a comment explaining this very large regex. It is not self-explanatory.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/DirectlyAddressableRestModel.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+/**
+ * A directly addressable REST resource
+ * 
+ * @author Andrea Bollini (andrea.bollini at 4science.it)
+ *
+ */
+public interface DirectlyAddressableRestModel extends RestModel {
```
Comment (tdonohue) [inline dspace-spring-rest/src/main/java/org/dspace/app/rest/model/DirectlyAddressableRestModel.java]: I'm not entirely sure I understand the meaning of "directly addressable" in this class name/description. I'm not sure I know of a better name for this interface, but it's currently a confusing name to me.

File: dspace/config/dspace.cfg [hunk 11]
```diff
@@ -28,8 +28,8 @@ dspace.hostname = localhost
 # DSpace base host URL.  Include port number etc.
 dspace.baseUrl = http://localhost:8080
 
-# The user interface you will be using for DSpace. Common usage is either xmlui or jspui
-dspace.ui = xmlui
+# The user interface you will be using for DSpace.
+dspace.ui = dspace-spring-rest
```
Comment (tdonohue) [inline dspace/config/dspace.cfg]: Was this change on purpose? I'm not against this conceptually, but we may wish to consider (eventually) renaming the config, as obviously the REST app is no longer a "ui".  Perhaps it should be `dspace.path` or even `dspace.rest.url` eventually (as presumably, we should support having REST running on a separate server from the Angular UI)

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/DirectlyAddressableRestModel.java (hunk 10 above)
Comment (abollini) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/model/DirectlyAddressableRestModel.java]: We need to distinguish between model that can be only embedded in other model such as the checksum (that have a value and an algorithm) and model that are exposed as Rest Resource (so they have an URL)

**Comment by tdonohue** [comment]
@abollini and @lap82 : I gave this a pretty thorough code review today (see inline comments above). Overall, it's looking good. But, there were some areas that I feel could be better commented and/or cleaned up slightly.  Nothing major though.

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/model/DirectlyAddressableRestModel.java (hunk 10 above)
Comment (tdonohue) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/model/DirectlyAddressableRestModel.java]: Could it just be called `RestResourceModel`?  And then we can add in a clearer comment to explain what we mean by that?  We could also ask around for other ideas...or maybe "directly addressable" makes sense to others (but it definitely is a confusing name for me).

**Comment by abollini** [comment]
hi @tdonohue 
thanks to review it! I will need to check with @lap82 why we have left in place some classes and methods. I know that they were here at the start because we were uncertain about how to deal with them in the new architecture and we have left to avoid the need to anticipate a refactoring in other part of the application without enough understanding... but they are likely to be removed now.

**Comment by abollini** [comment]
replaced by #1929
